src/QsCompiler/Transformations/CallGraph/SimpleCallGraph.cs

```diff
@@ -84,7 +84,7 @@ private static IEnumerable<IEnumerable<T>> CartesianProduct<T>(IEnumerable<IEnum
             IEnumerable<IEnumerable<T>> result = new[] { Enumerable.Empty<T>() };
             foreach (var sequence in sequences)
             {
-                result = sequence.SelectMany(item => result, (item, seq) => seq.Concat(new[] { item }));
+                result = sequence.SelectMany(item => result, (item, seq) => seq.Concat(new[] { item })).ToList();
```

Comment:
Why is `ToList()` needed?

Reply:
The SelectMany is lazy, and since the thing being assigned, `result`, is used in the selector as well, the code get confused and fails if `result` is not forced to be evaluated after each iteration.

Reply:
Could you add a comment about that since it doesn't seem obvious to me from the code?

Reply:
done